### PR TITLE
Hypersphere loss

### PR DESCRIPTION
**Pull request by EelcoHoogendoorn**
Pr for https://github.com/lightly-ai/lightly/issues/250

Played around a lot with all kinds of different variations to this type of loss function, but decided it was getting out of hand, so I just left this PR with the literal implementation as provided in the published paper.

I dont like the temperature parameter this method has, its similar to the one found in the crossentropy; nor do I like the weight between the two losses. In practice I find that the behavior of the loss is data dependent. I have observed funny behavior in the past some for types of data and architectures. Ideally, the average distance between negative samples tends to sqrt(2), which is its maximum on a hypersphere of any dimension. Often this is approximated, but this certainly isnt guaranteed without tweaking. However, as provided this loss does (EDIT: said 'not' here but that was a typo) seem quite robust to catastropic collapse.

Also, ive experimented with various ways to make the 'temperature' parameter trainable, or scaling with the data distribution. This is quite important to convergence and quality; but its a huge rabbit hole, so yeah ill just stick to copying the paper here.

### CODE REVIEW

File: lightly/loss/hypersphere_loss.py
Comment (IgorSusmelj) [inline lightly/loss/hypersphere_loss.py:11]: Can we use the PyTorch functional `normalize` for this?
(see https://pytorch.org/docs/1.1.0/nn.html#normalize)

```python
import torch.nn.functional as F
x_normalized = F.normalize(x)
```

File: lightly/loss/hypersphere_loss.py
Comment (IgorSusmelj) [inline lightly/loss/hypersphere_loss.py:85]: I would leave it here since it's needed for the loss calculation. For inference, I would leave it to the user to do the normalization.

File: lightly/loss/hypersphere_loss.py
Comment (IgorSusmelj) [inline lightly/loss/hypersphere_loss.py:95]: You can drop the `nan_to_num` here. As you mentioned, the case of batch size 1 makes not much sense. There is no need to make the tests pass that case. 
`nan_to_num` would additionally require us to switch the minimum PyTorch version to 1.8.

**Review by IgorSusmelj — CHANGES_REQUESTED** [review]
Great work! Thank you very much for the contribution. I tested the code and it works :)

I will do some benchmarking and a more in-depth analysis of the new loss later.

Would do some minor modifications to the code before merging:
- Use PyTorch built-in normalize function
- No need to support batch size 1 for the test case

File: lightly/loss/hypersphere_loss.py
Comment (EelcoHoogendoorn) [reply lightly/loss/hypersphere_loss.py:95]: Alright; will do. Wasn't keen on the overhead of another call in the graph for purely theoretical reasons either.

File: lightly/loss/hypersphere_loss.py
Comment (EelcoHoogendoorn) [reply lightly/loss/hypersphere_loss.py:85]: Alright; will make sure the docstring calls this out as explicitly as possible.

File: lightly/loss/hypersphere_loss.py
Comment (EelcoHoogendoorn) [reply lightly/loss/hypersphere_loss.py:11]: Seems like it. Implementing it myself is a reflex of numpy refusing to do so for the part 10 years ive been using it I suppose.

**Comment by EelcoHoogendoorn** [comment]
Adressed the above comments; the only open question I have is about the naming of things; not sure that Hypersphere loss really captures the essence of it... but its not like I have better ideas.

**Review by IgorSusmelj — APPROVED** [review]

**Pull request merged**
